query/common.d.ts

```diff
@@ -0,0 +1,278 @@
+   * The required order for the result stream.
+   */
+  order?: QueryOperationOrder<OrderItemsType>;
+
+  /**
+   * Custom properties
+   */
+  [key: string]: any;
+}
+
+/**
+ * Generic interface that defines query objects following the Future pattern.
+ */
+export interface BaseQuery {
+  /**
+   * Identifier for the type of result of tis query.
+   */
+  resultType: string;
+
+  /**
+   * Returns either a stream containing all the items that match the given query,
+   * a boolean or void depending on the semantics of the given query.
+   */
+  execute(opts?: any): Promise<ResultStream<any> | boolean | void>;
+}
+
+/**
+ * Query object that returns bindings.
+ */
+export interface QueryBindings extends BaseQuery, BaseMetadataQuery<RDF.Variable, { variables: RDF.Variable[] }> {
+  resultType: 'bindings';
+  execute(opts?: QueryExecuteOptions<RDF.Variable>): Promise<ResultStream<Bindings>>;
+}
+
+/**
+ * Query object that returns quads.
+ */
+export interface QueryQuads extends BaseQuery, BaseMetadataQuery<QuadTermName, unknown> {
+  resultType: 'quads';
+  execute(opts?: QueryExecuteOptions<QuadTermName>): Promise<ResultStream<RDF.Quad>>;
+}
+
+/**
+ * Query object that returns a boolean.
+ */
+export interface QueryBoolean extends BaseQuery {
+  resultType: 'boolean';
+  execute(): Promise<boolean>;
+}
+
+/**
+ * Query object that returns void.
+ */
+export interface QueryVoid extends BaseQuery {
+  resultType: 'void';
+  execute(): Promise<void>;
+}
+
+/**
+ * Union type for the different query types.
+ */
+export type Query = QueryBindings | QueryBoolean | QueryQuads | QueryVoid;
+
+/**
+ * Bindings represents the mapping of variables to RDF values using an immutable Map-like representation.
+ * This means that methods such as `set` and `delete` do not modify this instance,
+ * but they return a new Bindings instance that contains the modification.
+ *
+ * Bindings instances are created using a BindingsFactory.
+ *
+ * The internal order of variable-value entries is undefined.
+ */
+export interface Bindings extends Iterable<[RDF.Variable, RDF.Term]> {
+  type: 'bindings';
+  /**
+   * Check if a binding exist for the given variable.
+   * @param key A variable
+   */
+  has: (key: RDF.Variable) => boolean;
+  /**
+   * Obtain the binding value for the given variable.
+   * @param key A variable
+   */
+  get: (key: RDF.Variable) => RDF.Term | undefined;
+  /**
+   * Create a new Bindings object by adding the given variable and value mapping.
+   *
+   * If the variable already exists in the binding, then the existing mapping is overwritten.
+   *
+   * @param key The variable key.
+   * @param value The value.
+   */
+  set: (key: RDF.Variable, value: RDF.Term) => Bindings;
+  /**
+   * Create a new Bindings object by removing the given variable.
+   *
+   * If the variable does not exist in the binding, a copy of the Bindings object is returned.
+   *
+   * @param key The variable key.
+   */
+  delete: (key: RDF.Variable) => Bindings;
```

Comment:
Moving modifiers at the instance level is something I had to do, too. +1 from me.